### PR TITLE
Update ir spec code to make use of streams, helpers and dag + modify server record code

### PR DESCRIPTION
**Pull request by anna-cross**
Covers tickets: 

https://github.com/meroxa/turbine-core/issues/59
https://github.com/meroxa/turbine-core/issues/49

Two sets of changes: 

IR - 
Added helps and dag creation in ir code. The helpers are used in server changes and dag is used to validate data flow (check no loops). 

Server - 
Makes use of the helper code

### CODE REVIEW

File: pkg/ir/spec.go [hunk 1]
```diff
@@ -22,11 +23,19 @@ const (
 )
 
 type DeploymentSpec struct {
-	Secrets    map[string]string `json:"secrets,omitempty"`
-	Connectors []ConnectorSpec   `json:"connectors"`
-	Functions  []FunctionSpec    `json:"functions,omitempty"`
-	Streams    []StreamSpec      `json:"streams,omitempty"`
-	Definition DefinitionSpec    `json:"definition"`
+	mu            sync.Mutex
+	DeploymentMap DeploymentMap
```
Comment (lyuboxa) [inline pkg/ir/spec.go:27]: the map doesn't have to be exported since it is internal information.

File: pkg/ir/spec.go [hunk 2]
```diff
@@ -98,3 +104,121 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
+	}
+
+	if source := d.turbineDAG.nodes[c.UUID]; source != "" {
+		return fmt.Errorf("connector with uuid %s already exists", c.UUID)
+	}
+	d.Connectors = append(d.Connectors, *c)
+	d.turbineDAG.source = c
+	if src, err := d.turbineDAG.dag.AddVertex(&c); err != nil {
+		return err
+	} else {
```
Comment (lyuboxa) [inline pkg/ir/spec.go:123]: what is the purpose of the else here?

File: pkg/ir/spec.go [hunk 3]
```diff
@@ -98,3 +104,121 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
+	return nil
+}
+
+func (d *DeploymentSpec) AddStream(s *StreamSpec) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	source := d.turbineDAG.nodes[s.FromUUID]
+	dest := d.turbineDAG.nodes[s.ToUUID]
+
+	if source == "" {
```
Comment (lyuboxa) [inline pkg/ir/spec.go:167]: ideally you wnat to check for existence not for value, so this is more appropriate:
```
source, ok := ....
if !ok {
  .. error
}
```

File: pkg/ir/spec.go [hunk 4]
```diff
@@ -98,3 +104,121 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
+		return fmt.Errorf("source node (%s) does not exist", s.FromUUID)
+	}
+
+	if dest == "" {
+		return fmt.Errorf("destination node (%s) does not exist", s.ToUUID)
+	}
+
+	if s.FromUUID == s.ToUUID {
```
Comment (lyuboxa) [inline pkg/ir/spec.go:175]: i think `AddEdge` may catch that, but sure.

File: pkg/ir/spec.go [hunk 5]
```diff
@@ -98,3 +104,121 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
+		return fmt.Errorf("for stream %q , ids for source (%q) and destination (%q) must be different.", s.Name, s.FromUUID, s.ToUUID)
+	}
+
+	d.Streams = append(d.Streams, *s)
+	if err := d.turbineDAG.dag.AddEdge(source, dest); err != nil {
+		return fmt.Errorf("cannot create an edge from  UUID %s to UUID %s, it will result in a loop", s.FromUUID, s.ToUUID)
```
Comment (lyuboxa) [inline pkg/ir/spec.go:181]: return the err as is.

File: pkg/ir/spec.go [hunk 6]
```diff
@@ -98,3 +104,121 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
+
+func (d *DeploymentSpec) InitDag() {
+	d.turbineDAG = turbineDAG{
+		nodes: make(map[string]string),
+		dag:   dag.NewDAG(),
+		//surce: d.turbineDAG.source,
+	}
+}
+
+func (d *DeploymentSpec) BuildDAG() error {
```
Comment (lyuboxa) [inline pkg/ir/spec.go:194]: i don't think you need this method.

File: pkg/ir/spec.go (hunk 2 above)
Comment (anna-cross) [reply pkg/ir/spec.go]: This can be simplified yup

File: pkg/server/record_test.go [hunk 7]
```diff
@@ -184,8 +165,8 @@ func TestReadCollection(t *testing.T) {
 				require.EqualError(t, err, test.errMsg)
 			} else {
 				require.Nil(t, err)
-				require.Equal(t, &pb.Collection{}, res)
-				require.Equal(t, test.want, s.deploymentSpec)
+				require.Equal(t, test.req.Resource.Collection, res.Name)
+				require.NotEmpty(t, s.deploymentSpec.Connectors)
```
Comment (dianadoherty) [inline pkg/server/record_test.go:169]: we should do a more thorough test here instead of just checking the connectors arent empty

File: pkg/server/record_test.go [hunk 8]
```diff
@@ -301,38 +298,43 @@ func TestWriteCollectionToResource(t *testing.T) {
...
+			Collection: read,
+		},
+	)
+
 	require.Nil(t, err)
-	require.Equal(t, &pb.Collection{}, res)
-	require.Equal(t, want, s.deploymentSpec)
+	require.NotEmpty(t, res)
+	require.NotEmpty(t, s.deploymentSpec.Functions)
+	require.NotEmpty(t, s.deploymentSpec.Streams)
```
Comment (dianadoherty) [inline pkg/server/record_test.go:331]: I think we need a more thorough check than just not empty

File: pkg/ir/spec.go [hunk 9]
```diff
@@ -22,13 +25,21 @@ const (
...
 	Secrets    map[string]string `json:"secrets,omitempty"`
 	Connectors []ConnectorSpec   `json:"connectors"`
 	Functions  []FunctionSpec    `json:"functions,omitempty"`
 	Streams    []StreamSpec      `json:"streams,omitempty"`
 	Definition DefinitionSpec    `json:"definition"`
 }
 
+type turbineDAG struct {
+	dag    *dag.DAG
+	source *ConnectorSpec
```
Comment (lyuboxa) [inline pkg/ir/spec.go:39]: come to think about this, maybe we don't need that, since `dag.GetRoots()` will provide that info based on the graph, what do you think?

File: pkg/server/record_test.go [hunk 10]
```diff
@@ -350,8 +352,6 @@ func TestRegisterSecret(t *testing.T) {
 		})
 	require.Nil(t, err)
 	require.Equal(t, empty(), res)
-
-	require.Equal(t, want, s.deploymentSpec)
```
Comment (dianadoherty) [inline pkg/server/record_test.go:354]: We should still keep this

**Review by dianadoherty — COMMENTED** [review]
Amazing job! I just have small test and error message nits

File: pkg/ir/spec.go [hunk 11]
```diff
@@ -22,13 +25,21 @@ const (
...
+	nodes  map[string]string
```
Comment (lyuboxa) [inline pkg/ir/spec.go:40]: Another question, do we need nodes now that we can do `dag.GetVertex()` and get the pointer to the connector spec?

File: pkg/ir/spec.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (lyuboxa) [reply pkg/ir/spec.go:39]: or not have it all in there, since the dag can tell us which is the root/source

Comment (anna-cross) [reply pkg/ir/spec.go:39]: Yup! It can go :)

[the remark this answers is not in the document]
Comment (anna-cross) [reply pkg/ir/spec.go:39]: Yup no need to track it anymore

File: pkg/ir/spec.go [hunk 12]
```diff
@@ -98,3 +104,92 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
+func (d *DeploymentSpec) AddSource(c *ConnectorSpec) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	// check if spec is source
+	if d.turbineDAG.source != nil {
+		return fmt.Errorf("source connector (%s) already exists", d.turbineDAG.source.UUID)
+	}
+
+	if source := d.turbineDAG.nodes[c.UUID]; source != "" {
+		return fmt.Errorf("connector with uuid %s already exists", c.UUID)
```
[the remark this answers is not in the document]
Comment (anna-cross) [reply pkg/ir/spec.go:117]: Yeah, to be fair we can get rid of this as it will never reach this point. 

It was mainly to catch an accidental dupe of a source but that is already caught by the error check above. 

Outside of this code, yes, there can only be one source node. If there are multiple destination nodes that are pulling from one source node, then we just have multiple streams for each connection.

File: pkg/ir/spec.go [hunk 13]
```diff
@@ -98,3 +104,92 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
+	}
+	d.turbineDAG.nodes[c.UUID] = src
+	return nil
+}
+
+func (d *DeploymentSpec) AddFunction(f *FunctionSpec) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if function := d.turbineDAG.nodes[f.UUID]; function != "" {
+		return fmt.Errorf("function with uuid %s already exists", f.UUID)
```
[the remark this answers is not in the document]
Comment (anna-cross) [reply pkg/ir/spec.go:133]: Yup this is not needed anymore either~

File: pkg/ir/spec.go (hunk 11 above)
Comment (anna-cross) [reply pkg/ir/spec.go:40]: We can use GetVertices if we want, we will just have to loop through each vertex to get the pointer in AddStream to check the spec's id against fromUUID and toUUID

File: pkg/ir/spec.go [hunk 14]
```diff
@@ -98,3 +103,81 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
+
+	if len(d.turbineDAG.dag.GetRoots()) >= 1 {
+		return fmt.Errorf("source connector already exists, can only add one per application")
+	}
+	d.Connectors = append(d.Connectors, *c)
+	src, err := d.turbineDAG.dag.AddVertex(&c)
+	if err != nil {
+		return err
+	}
+	d.turbineDAG.nodes[c.UUID] = src
```
Comment (lyuboxa) [inline pkg/ir/spec.go:119]: I seriously think you don't need this, since `d.turbineDAG.dag.GetVertex(c.UUID)` will give you the ptr of that object?

so we won't need nodes at all.

File: pkg/ir/spec.go [hunk 15]
```diff
@@ -22,13 +25,20 @@ const (
 )
 
 type DeploymentSpec struct {
+	mu         sync.Mutex
+	turbineDAG turbineDAG
```
Comment (lyuboxa) [inline pkg/ir/spec.go:29]:
```suggestion
	turbineDAG *dag.DAG
```

File: pkg/ir/spec.go [hunk 16]
```diff
@@ -22,13 +25,19 @@ const (
...
 	Secrets    map[string]string `json:"secrets,omitempty"`
 	Connectors []ConnectorSpec   `json:"connectors"`
 	Functions  []FunctionSpec    `json:"functions,omitempty"`
 	Streams    []StreamSpec      `json:"streams,omitempty"`
 	Definition DefinitionSpec    `json:"definition"`
 }
 
+type turbineDAG struct {
```
Comment (lyuboxa) [inline pkg/ir/spec.go:37]: do we need a separate struct for that?

File: pkg/ir/spec.go [hunk 17]
```diff
@@ -98,3 +98,66 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
+
+func (d *DeploymentSpec) AddSource(c *ConnectorSpec) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if len(d.turbineDAG.dag.GetRoots()) >= 1 {
+		return fmt.Errorf("source connector already exists, can only add one per application")
+	}
+	d.Connectors = append(d.Connectors, *c)
+	err := d.turbineDAG.dag.AddVertexByID(c.UUID, &c)
```
Comment (lyuboxa) [inline pkg/ir/spec.go:110]:
```suggestion
	if err := d.turbineDAG.dag.AddVertexByID(c.UUID, &c); err != nil {
```

File: pkg/ir/spec_test.go [hunk 18]
```diff
@@ -320,3 +184,759 @@ func Test_MarshalUnmarshal(t *testing.T) {
...
+			},
+		},
+	)
+	require.EqualError(t, err, fmt.Errorf("source connector already exists, can only add one per application").Error())
+}
+
+// Scenario 1 - Simple DAG
+// source → fn -> dest
+// ( src_con ) → (stream) → (function) → (stream) → (dest1)
+func Test_Scenario1(t *testing.T) {
```
Comment (dianadoherty) [inline pkg/ir/spec_test.go:220]: What happens if you create a disconnected graph? Would it error if we have something like this? 
```
 src -> fn[0]

fn[1] -> dst
```

File: pkg/ir/spec.go (hunk 16 above)
Comment (dianadoherty) [reply pkg/ir/spec.go:37]: Isn't that what you commented/recommended? https://github.com/meroxa/turbine-core/pull/63#discussion_r1047805844

File: pkg/ir/spec.go [hunk 19]
```diff
@@ -98,3 +98,66 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
+	}
+
+	d.Streams = append(d.Streams, *s)
+	if err := d.turbineDAG.dag.AddEdge(s.FromUUID, s.ToUUID); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *DeploymentSpec) InitDag() {
```
[the remark this answers is not in the document]
Comment (lyuboxa) [reply pkg/ir/spec.go:159]: you will need to modify the spec so it has this
```
type DeploymentSpec struct {
   mu sync.Mutex
   turbineDag dag.DAG
   dagInitOnce sync.Once

   ....
}
```

and 

```
func (d *DeploymentSpec) AddStream(s *StreamSpec) error {
	d.mu.Lock()
	defer d.mu.Unlock()

        d.init()
        
       ...
```

File: pkg/ir/spec.go (hunk 16 above)
Comment (lyuboxa) [reply pkg/ir/spec.go:37]: no, my suggestion recommended putting dag.DAG straight in the spec?

File: pkg/ir/spec.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (lyuboxa) [reply pkg/ir/spec.go:110]: the dag should definite

File: pkg/ir/spec.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (anna-cross) [reply pkg/ir/spec.go:159]: Had no idea you could do this! This is great :D

File: pkg/ir/spec_test.go (hunk 18 above)
Comment (anna-cross) [reply pkg/ir/spec_test.go:220]: Uh oh- looks like this is a breaking case (should be throwing an error but it doesnt) 

Looking into it

File: pkg/ir/spec.go (hunk 16 above)
Comment (dianadoherty) [reply pkg/ir/spec.go:37]: Oh right, my code reading comprehension was off. Ignore me 🙈

File: pkg/ir/spec.go [hunk 20]
```diff
@@ -102,23 +99,25 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
-	if len(d.turbineDAG.dag.GetRoots()) >= 1 {
+	if len(d.turbineDag.GetRoots()) >= 1 {
 		return fmt.Errorf("source connector already exists, can only add one per application")
 	}
-	d.Connectors = append(d.Connectors, *c)
-	err := d.turbineDAG.dag.AddVertexByID(c.UUID, &c)
-	if err != nil {
-		return err
+	if c.Type != ConnectorSource {
+		return fmt.Errorf("connector type isn't a destination, please check you are writing to destination connector.")
```
Comment (dianadoherty) [inline pkg/ir/spec.go]:
```suggestion
		return fmt.Errorf("connector type isn't a source, please check you are reading from source connector.")
```

File: pkg/ir/spec.go [hunk 21]
```diff
@@ -102,23 +99,25 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
 	d.mu.Lock()
 	defer d.mu.Unlock()
+	d.init()
+
 	d.Functions = append(d.Functions, *f)
-	err := d.turbineDAG.dag.AddVertexByID(f.UUID, &f)
+	err := d.turbineDag.AddVertexByID(f.UUID, &f)
 	if err != nil {
 		return err
 	}
```
Comment (dianadoherty) [inline pkg/ir/spec.go:123]:
```suggestion
	reutrn d.turbineDag.AddVertexByID(f.UUID, &f)
```

File: pkg/ir/spec.go [hunk 22]
```diff
@@ -25,17 +25,14 @@ const (
...
-	Connectors []ConnectorSpec   `json:"connectors"`
-	Functions  []FunctionSpec    `json:"functions,omitempty"`
-	Streams    []StreamSpec      `json:"streams,omitempty"`
-	Definition DefinitionSpec    `json:"definition"`
-}
-
-type turbineDAG struct {
-	dag *dag.DAG
+	mu          sync.Mutex
+	turbineDag  dag.DAG
```
Comment (dianadoherty) [inline pkg/ir/spec.go:29]:
```suggestion
	turbineDag  *dag.DAG
```

File: pkg/ir/spec.go [hunk 23]
```diff
@@ -98,3 +95,80 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
+func (d *DeploymentSpec) AddSource(c *ConnectorSpec) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.init()
+
+	if len(d.turbineDag.GetRoots()) >= 1 {
+		return fmt.Errorf("source connector already exists, can only add one per application")
+	}
+	if c.Type != ConnectorSource {
+		return fmt.Errorf("connector type isn't a source, please check you are reading from a source connector.")
```
Comment (lyuboxa) [inline pkg/ir/spec.go:108]: these errors can be less wordy,  consider something like
```
not a source connector
```

File: pkg/ir/spec.go [hunk 24]
```diff
@@ -98,3 +95,80 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
 	return spec, nil
 }
+
+func (d *DeploymentSpec) AddSource(c *ConnectorSpec) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.init()
+
+	if len(d.turbineDag.GetRoots()) >= 1 {
+		return fmt.Errorf("source connector already exists, can only add one per application")
```
Comment (lyuboxa) [inline pkg/ir/spec.go:105]: same here, more succinct `too many source connectors`

File: pkg/ir/spec.go [hunk 25]
```diff
@@ -98,3 +95,80 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
+	return nil
+}
+
+func (d *DeploymentSpec) AddDestination(c *ConnectorSpec) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.init()
+
+	if c.Type != ConnectorDestination {
+		return fmt.Errorf("connector type isn't a destination, please check you are writing to destination connector.")
```
Comment (lyuboxa) [inline pkg/ir/spec.go:133]: ditto re messaging in previous comment

File: pkg/ir/spec.go [hunk 26]
```diff
@@ -98,3 +95,80 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
+	return d.turbineDag.AddVertexByID(c.UUID, &c)
+}
+
+func (d *DeploymentSpec) AddStream(s *StreamSpec) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.init()
+
+	if _, err := d.turbineDag.GetVertex(s.FromUUID); err != nil {
+		return fmt.Errorf("source with UUID - (%s) does not exist", s.FromUUID)
```
Comment (lyuboxa) [inline pkg/ir/spec.go:145]: similarly `source %q does not exist`

File: pkg/ir/spec.go [hunk 27]
```diff
@@ -98,3 +95,80 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
+	}
+
+	if _, err := d.turbineDag.GetVertex(s.ToUUID); err != nil {
+		return fmt.Errorf("destination with UUID - (%s) does not exist", s.ToUUID)
+	}
+
+	d.Streams = append(d.Streams, *s)
+	if err := d.turbineDag.AddEdge(s.FromUUID, s.ToUUID); err != nil {
```
Comment (lyuboxa) [inline pkg/ir/spec.go:153]: perhaps

```suggestion
	return d.turbineDag.AddEdge(s.FromUUID, s.ToUUID)
```

File: pkg/ir/spec.go [hunk 28]
```diff
@@ -98,3 +95,80 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
+		return err
+	}
+	return nil
+}
+
+func (d *DeploymentSpec) ValidateDag() error {
```
Comment (lyuboxa) [inline pkg/ir/spec.go:159]:
```suggestion
func (d *DeploymentSpec) ValidateDAG() error {
```

File: pkg/ir/spec.go [hunk 29]
```diff
@@ -98,3 +95,80 @@ func Unmarshal(data []byte) (*DeploymentSpec, error) {
...
+	if len(d.turbineDag.GetRoots()) > 1 {
+		return fmt.Errorf("more than one source / root detected, can only add one per application. please ensure your resources are connected.")
+	}
+	return nil
+}
+
+func (d *DeploymentSpec) BuildDAG() string {
+	return d.turbineDag.String()
```
Comment (lyuboxa) [inline pkg/ir/spec.go:167]: This should be returning `d.turbineDAG` so that it can be used by platform, more importantly,
it should ensure the `turbineDAG` is not empty. Considering this case in platform:

```
spec, err := ir.Unmarshal(&b)
if err != nil {
  return err
}

// using the dag to walk the graph and provision all vertices 
dag := spec.BuildDAG()
// provision source connector
sourceSpec := dag.GetRoots()[0]
// provision sourceSpec  ...

```

Comment (lyuboxa) [reply pkg/ir/spec.go:167]: I think this PR needs to be shipped, so we can address the build when platform work starts. So preferably just return the dag for now.

**Review by lyuboxa — APPROVED** [review]
Looks good, left some comments

File: pkg/ir/spec.go [hunk 30]
```diff
@@ -21,12 +24,18 @@ const (
 	LatestSpecVersion = "0.2.0"
 )
 
+type idInterface interface {
+	ID() string
+}
```
Comment (dianadoherty) [inline pkg/ir/spec.go:29]: I don't think this is used anywhere

**Review by dianadoherty — APPROVED** [review]

**Review by dianadoherty — APPROVED** [review]

File: pkg/ir/spec.go [hunk 31]
```diff
@@ -142,33 +144,58 @@ func (d *DeploymentSpec) AddStream(s *StreamSpec) error {
...
+func (d *DeploymentSpec) BuildDAG() (*dag.DAG, error) {
+	turbineDag := dag.NewDAG()
+	for _, con := range d.Connectors {
+		cont := &con
+		if err := turbineDag.AddVertexByID(con.UUID, &cont); err != nil {
+			return nil, err
+		}
+	}
+	for _, fun := range d.Functions {
+		funn := &fun
```
Comment (lyuboxa) [inline pkg/ir/spec.go:181]: why are you doing this?

Comment (anna-cross) [reply pkg/ir/spec.go:181]: The  AddVertexByID() wants a pointer of a pointer passed to it, or else it starts freaking out about the hash thing

Comment (anna-cross) [reply pkg/ir/spec.go:181]: Oh oops, yup I see your edited comment, that is a better way

**Review by dianadoherty — APPROVED** [review]
